File: drools/drools-examples/src/main/org/drools/examples/decisiontables/TeamAllocationExample.java
/*
 * Created on 2/06/2005
 */
package org.drools.examples.decisiontables;



import java.io.IOException;
import java.io.InputStream;
import java.util.Calendar;

import org.drools.FactException;
import org.drools.IntegrationException;
import org.drools.RuleBase;
import org.drools.WorkingMemory;
import org.drools.decisiontable.DecisionTableLoader;
import org.drools.examples.decisiontables.model.Claim;
import org.drools.examples.decisiontables.model.Team;
import org.xml.sax.SAXException;

/**
 * @author <a href="mailto:michael.neale@gmail.com"> Michael Neale</a>
 *
 * Show off a more complex example.
 * Runs a few different scenarios through the same rule base, but only building the rule base once.
 * 
 */
public class TeamAllocationExample {
	
	private RuleBase _ruleBase;
	

	
	private void loadRuleBase() throws SAXException, IOException, IntegrationException {
		InputStream stream = this.getClass().getResourceAsStream("TeamAllocationExample.xls");
		//MN this is all you have to do to get decision tables working...
		_ruleBase = DecisionTableLoader.loadFromInputStream(stream);
	}


	public static void main(String[] args) throws Exception {
		
		TeamAllocationExample example = new TeamAllocationExample();
		example.loadRuleBase();	
		example.teamAllocationBasic();
		example.teamAllocationLongtail1();
		example.teamAllocationLongtail2();
	}	
	
	
	public void teamAllocationBasic() throws Exception {
				
		Claim cat = getCatastrophic();
		Team team = new Team();
		
		executeRules(cat, team);
		System.out.println(team.getName());
	}
	
	public void teamAllocationLongtail1() throws Exception {
		Claim claim = getBetweenSomeDateLongtail1();
		Team team = new Team();
				
		executeRules(claim, team);		
		System.out.println(team.getName());
	}
	
	public void teamAllocationLongtail2() throws Exception {
		Claim claim = getBetweenSomeDateLongtail2();
		Team team = new Team();

				
		executeRules(claim, team);
		
		System.out.println(team.getName());

	}
	
	

	private void executeRules(Claim claim, Team team) throws SAXException, IOException, IntegrationException, FactException {
		WorkingMemory engine = _ruleBase.newWorkingMemory();
		engine.assertObject(claim);
		engine.assertObject(team);
		engine.fireAllRules();
	}
	
	
	//set up the example scenarios below
	private Claim getBetweenSomeDateLongtail1() {
		Claim claim = new Claim();
		claim.setCatastrophic(false);
		Calendar cal = Calendar.getInstance();
		cal.set(2001, 1, 4);
		claim.setDateOfAccident(cal.getTime());
		claim.setInsuranceClass("1");
		claim.setAllocationCode("S1");
		claim.setClaimType("C");
		claim.setInsuredVehicleOwner("Personal");
		return claim;
		
	}
	
	private Claim getBetweenSomeDateLongtail2() {
		Claim claim = new Claim();
		claim.setCatastrophic(false);
		Calendar cal = Calendar.getInstance();
		cal.set(2002, 1, 4);
		claim.setDateOfAccident(cal.getTime());
		claim.setInsuranceClass("1");
		claim.setAllocationCode("S1");
		claim.setClaimType("C");
		claim.setInsuredVehicleOwner("Personal");
		return claim;
		
	}	

	private Claim getCatastrophic() {
		Claim claim = new Claim();
		claim.setCatastrophic(true);
		claim.setDateOfAccident(Calendar.getInstance().getTime());
		claim.setInsuranceClass("1");
		claim.setAllocationCode("S1");
		claim.setClaimType("1");
		claim.setInsuredVehicleOwner("Personal");
		return claim;
	}
	
}

File: trunk/drools/drools-dotnet/src/drools.dotnet.examples/decisiontables/TeamAllocationExample.cs
using System;
using Claim = org.drools.dotnet.examples.decisiontables.model.Claim;
using Team = org.drools.dotnet.examples.decisiontables.model.Team;
using NUnit.Framework;
using org.drools.dotnet.io;
namespace org.drools.dotnet.examples.decisiontables
{

    [TestFixture]
	public class TeamAllocationExample
	{
		private Claim BetweenSomeDateLongtail1
		{
			//set up the example scenarios below
			
			get
			{
				Claim claim = new Claim();
				claim.Catastrophic = false;
				System.Globalization.Calendar cal = new System.Globalization.GregorianCalendar();
				claim.DateOfAccident = cal.ToDateTime(2001, 1, 4, 0, 0, 0, 0, 0);
				claim.InsuranceClass = "1";
				claim.AllocationCode = "S1";
				claim.ClaimType = "C";
				claim.InsuredVehicleOwner = "Personal";
				return claim;

			}
			
		}
		private Claim BetweenSomeDateLongtail2
		{
			get
			{
				Claim claim = new Claim();
				claim.Catastrophic = false;
				System.Globalization.Calendar cal = new System.Globalization.GregorianCalendar();
				claim.DateOfAccident = cal.ToDateTime(2002, 1, 4, 0, 0, 0, 0, 0);
				claim.InsuranceClass = "1";
				claim.AllocationCode = "S1";
				claim.ClaimType = "C";
				claim.InsuredVehicleOwner = "Personal";
				return claim;
			}
			
		}
		private Claim Catastrophic
		{
			get
			{
				Claim claim = new Claim();
				claim.Catastrophic = true;
				claim.DateOfAccident = DateTime.Now;
				claim.InsuranceClass = "1";
				claim.AllocationCode = "S1";
				claim.ClaimType = "1";
				claim.InsuredVehicleOwner = "Personal";
				return claim;
			}
			
		}
		
		private RuleBase _ruleBase;
		
		
		
		private void  loadRuleBase()
		{
            _ruleBase = org.drools.dotnet.io.DecisionTableLoader.LoadFromStream(new System.IO.FileStream(System.Environment.CurrentDirectory + "\\drls\\TeamAllocationExample.xls", System.IO.FileMode.Open));
		}
		
		
		[Test]
		public void TestTeamAllocationExample()
        {
			
			
			loadRuleBase();
			teamAllocationBasic();
			teamAllocationLongtail1();
			teamAllocationLongtail2();
		}
		
		
		private void  teamAllocationBasic()
		{
			
			Claim cat = Catastrophic;
			Team team = new Team();
			
			executeRules(cat, team);
            Assert.AreEqual("Team Red", team.Name);
			System.Console.Out.WriteLine(team.Name);
		}
		
		private void  teamAllocationLongtail1()
		{
			Claim claim = BetweenSomeDateLongtail1;
			Team team = new Team();
			
			executeRules(claim, team);
            Assert.AreEqual("Longtail-1", team.Name);
			System.Console.Out.WriteLine(team.Name);
		}
		
		private void  teamAllocationLongtail2()
		{
			Claim claim = BetweenSomeDateLongtail2;
			Team team = new Team();
			
			
			executeRules(claim, team);
            Assert.AreEqual("Longtail-2", team.Name);
			System.Console.Out.WriteLine(team.Name);
		}
		
		
		
		private void  executeRules(Claim claim, Team team)
		{
			WorkingMemory engine = _ruleBase.GetNewWorkingMemory();
            
			engine.AssertObject(claim);
			engine.AssertObject(team);
			engine.FireAllRules();
		}
	}
}